Implement random unitary matrices

orlitzky opened the pull request:
Add a new `random_unitary_matrix()` function that is called from `matrix.random(..., algorithm='unitary')`.

orlitzky committed:
src/sage/matrix/special.py: add some "# long time" tags

I'm getting a few warnings about tests in this file that take longer
than 5s to run (without --long). Here we add "# long time" tags.

orlitzky committed:
src/doc/en/reference/references: add the LieOs1991 reference

New reference (Liebeck and Osborne, 1991) to be used in a new method
for generating random unitary matrices.

orlitzky committed:
src/sage/matrix/special.py: new random_unitary_matrix() function

Add a new function to generate random unitary matrices based on the
paper "The Generation of All Rational Orthogonal Matrices" by Liebeck
and Osborne (1991).

JohnCremona commented on src/sage/matrix/special.py:
```diff
@@ -3056,6 +3062,166 @@ def random_unimodular_matrix(parent, upper_bound=None, max_tries=100):
+    S = A - A.conjugate_transpose()
+    U = (S-I).inverse()*(S+I)
+    D = identity_matrix(F,n)
+    for i in range(n):
```
Why create D, you could reuse I?

orlitzky replied:
(This code has been sitting in my personal library for a while.) I was probably in a mood where I considered it a sin to have a variable named `I` that was no longer the identity matrix.

Regardless, I've replaced the diagonal `\pm 1` matrix by a loop that uses `U.set_row_to_multiple_of_row`, so now it's irrelevant. The new method is written in cython so it should be faster anyway. Thanks for the suggestion.

JohnCremona commented on src/sage/matrix/special.py:
```diff
@@ -3056,6 +3062,166 @@ def random_unimodular_matrix(parent, upper_bound=None, max_tries=100):
+    U = (S-I).inverse()*(S+I)
+    D = identity_matrix(F,n)
+    for i in range(n):
+        if ZZ.random_element(2).is_zero():
```
Is this really the best way to toss a coin?  random()<0.5 would be as good?

orlitzky replied:
Probably an irrational fear that my 50/50 probability will be wrong by the unit roundoff. I've replaced it. `random()` will definitely be faster.

JohnCremona commented on src/sage/matrix/special.py:
```diff
@@ -3056,6 +3062,166 @@ def random_unimodular_matrix(parent, upper_bound=None, max_tries=100):
+    D = identity_matrix(F,n)
+    for i in range(n):
+        if ZZ.random_element(2).is_zero():
+            D[i,i] *= F(-1)
```
It seems crazy that field elements are not required to have a negate() mthod, but apparently not.  But they do have a method called __neg__() with double underscores for evaluatio of unary minus.

I don't think it is necessary to construct F(-1) every time through the loop.  Perhaps  D[i,i] = -D[i,i] is better?

orlitzky replied:
This is handled within `set_row_to_multiple_of_row()` now (it casts the -1 to a rational for the flint API).

JohnCremona reviewed (changes requested):
This looks good.  I made some trivial coding comments.  Can you briefly say why you randomly negate rows (and is this the best way)?

orlitzky committed:
src/sage/matrix/special.py: optimize random_unitary_matrix()

Use the lower-level methods Matrix.set_row_to_multiple_of_row() and
random.random() to implement random_unitary_matrix(). This should
bring with it a small speed increase, and actually makes the code
simpler. Thanks to John Cremona for the suggestion.
